### PR TITLE
SYCL: Add a new header file with the reduction class definition

### PR DESCRIPTION
**Pull request by v-klochkov**
The patch only adds the new header file, it does not implement
the reduction functionality in parallel_for() yet.

Signed-off-by: Vyacheslav N Klochkov <vyacheslav.n.klochkov@intel.com>

### CODE REVIEW

**Comment by v-klochkov** [comment]
Sorry, it took several attempts for me to make 'clang-format' happy.

**Comment by v-klochkov** [comment]
> Since the `reducer` and `reduction` classes are Intel extensions, they need to be declared in `cl::sycl::intel`.
> 
> The files should be moved to CL/sycl/intel/ as well.

I moved the file to sycl/intel folder; put the 'reduction' function into namespace intel.
In my opinion, the aux classes 'reduction_impl' and 'reducer' should stay in 'detail' namespace as they are internals that are not supposed to be given to user, right?

@bader , Alexey what is your opinion on this?

**Comment by v-klochkov** [comment]
> Could you please add tests?

Ok, I added a very basic test.

**Comment by bader** [comment]
> > Since the `reducer` and `reduction` classes are Intel extensions, they need to be declared in `cl::sycl::intel`.
> > The files should be moved to CL/sycl/intel/ as well.
> 
> I moved the file to sycl/intel folder; put the 'reduction' function into namespace intel.
> In my opinion, the aux classes 'reduction_impl' and 'reducer' should stay in 'detail' namespace as they are internals that are not supposed to be given to user, right?
> 
> @bader , Alexey what is your opinion?

You can use `sycl::intel::detail` namespace for the symbols required for intel extensions by not supposed to be exposed to users.

File: sycl/test/reduction/reduction_ctor.cpp [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+  queue Q;
+  Q.submit([&](handler &CGH) {
+    accessor<T, Dim, access::mode::discard_write, access::target::global_buffer>
+        ReduAcc(ReduBuf, CGH);
+    auto Redu = intel::reduction(ReduAcc, BOp);
+    assert(Redu.getIdentity() == Identity &&
+           "Failed getIdentity() check().");
+    test_reducer(Redu, A, B);
+    test_reducer(Redu, Identity, A, B);
```
Comment (romanovvlad) [inline sycl/test/reduction/reduction_ctor.cpp:115]: Reading/writing from/to a device accessor should be illegal outside of kernel.
Is it possible to move this code to the kernel?

Comment (v-klochkov) [reply sycl/test/reduction/reduction_ctor.cpp:115]: Wait, this code does not read from the accessor. This code creates 'reducer' on host and reads on host. 'reducer' does not depend on CGH and device.

Comment (v-klochkov) [reply sycl/test/reduction/reduction_ctor.cpp:115]: Redu.Identity is also a host thing.

I'll create better tests for device when parallel_for is ready for review.

**Comment by v-klochkov** [comment]
> You can use `sycl::intel::detail` namespace for the symbols required for intel extensions by not supposed to be exposed to users.

Ok, I put aux reduction.hpp classes to cl::sycl::intel::detail namespace.
Please re-review.

File: sycl/test/reduction/reduction_ctor.cpp (hunk 1 above)
Comment (romanovvlad) [reply sycl/test/reduction/reduction_ctor.cpp:115]: Ok, please, add a comment that accessor is not used for reading and writing.

Comment (v-klochkov) [reply sycl/test/reduction/reduction_ctor.cpp:117]: Done. Thank you!

**Comment by v-klochkov** [comment]
Thank you for the approvals.
I see some LIT tests (that seem unrelated to my changes) fail on Windows (buildbot/sycl-win-x64-pr).
Still need to check/confirm that.

**Comment by v-klochkov** [comment]
The new cl::sycl::intel::detail namespace was added in the 1st commit.
That caused problems to compiler on Windows. For expressions like detail::something (used inside cl::sycl::intel) looked for "something" only in cl::sycl::intel::detail and did not look for it in cl::sycl::detail.
The fix changes "detail::something" to "sycl::detail::something".

**Pull request merged**